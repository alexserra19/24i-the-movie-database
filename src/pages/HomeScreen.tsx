import React, { useEffect, useState } from 'react';
import { View, StyleSheet, SafeAreaView, TouchableOpacity, Text, TextInput, Platform, Alert, Button, Image, Dimensions, ScrollView, Picker, FlatList } from 'react-native';
import AppConstants from '../utils/AppConstants';
import { ListItem, normalize, SearchBar } from 'react-native-elements';
import { useDispatch, useSelector } from "react-redux";
import { IsLoadingHoc } from "../components/HOCS/IsLoadingHOC";
import * as RootNavigation from "../navigation/RootNavigation";
import mediaActions from '../store/actions/mediaActions';
import _ from "lodash"
import { commonStyles } from '../styles/common';
import { Category, Media } from '../utils/typings';
import helpers from '../utils/helpers';
import Carousel from 'react-native-snap-carousel';
import SelectDropdown from 'react-native-select-dropdown'
import Icon from 'react-native-vector-icons/FontAwesome';
import { CarouselItem } from '../components/CarouselItem/CarouselItem';
import { MediaListItem } from '../components/MediaListItem/MediaListItem';
import RNPickerSelect from 'react-native-picker-select';

interface IHomeScreenProps {
    navigation: any;
    setLoading: Function
}

const HomeScreen = (props: IHomeScreenProps) => {
    const [search, updateSearch] = useState<string>('');
    const movies = useSelector((store: any) => store.mediaReducer.movies);
    const tvSeries = useSelector((store: any) => store.mediaReducer.tvSeries);
    const moviesCategories = useSelector((store: any) => store.mediaReducer.moviesCategories);
    const tvSeriesCategories = useSelector((store: any) => store.mediaReducer.tvSeriesCategories);
    const [screenWidth, updateScreenWidth] = useState<number>(Dimensions.get('window').width);
    const [movieCategorySelected, setMovieCategorySelected] = useState(null);
    const [tvSerieCategorySelected, setTvSerieCategorySelected] = useState(null);
    const [isLandscape, setIsLandscape] = useState<Boolean>(helpers.isLandscape());

    const dispatch = useDispatch()

    useEffect(() => {
        props.setLoading(true)
        dispatch(mediaActions.initializeStart())

        Dimensions.addEventListener('change', () => {
            updateScreenWidth(Dimensions.get('window').width)
            setIsLandscape(helpers.isLandscape())
        })

        return () => {
            Dimensions.removeEventListener('change', () => {
                updateScreenWidth(Dimensions.get('window').width)
                setIsLandscape(helpers.isLandscape())
            })
        }
    }, []);


    useEffect(() => {
        if (!_.isEmpty(movies) && !_.isEmpty(tvSeries) && !_.isEmpty(moviesCategories)) {
            props.setLoading(false)
        }
    }, [movies, tvSeries]);


    const doSearch = (text: string) => {
        updateSearch(text)
    }


    const renderPrincipalMovies = () => {
        const isLandscape = helpers.isLandscape()
        return (
            <Carousel
                data={movies}
                renderItem={renderItemPrincipalMovie}
                sliderWidth={screenWidth - normalize(isLandscape ? 110 : 20)}
                itemWidth={screenWidth - normalize(100)}
            />
        )
    }

    const viewItemDetails = (item: Media) => {
        dispatch(mediaActions.selectMedia(item))
        RootNavigation.navigate(AppConstants.routeName.details)
    }

    const renderItemPrincipalMovie = ({ item }: any) => {
        return (
            <CarouselItem
                item={item}
                onPress={viewItemDetails}
                isLandscape={isLandscape}
            />
        );
    }

    const renderCategoryItems = (
        categoriesList: Array<Category>,
        title: string,
        updateCategorySelected: Function,
        mediaList: Array<Media>,
        categorySelected: number
    ) => {
        let selectItems = categoriesList.map((item) => ({ label: item.genre, value: item.id }))
        return (
            <View>
                <View style={styles.categoriesContainer}>
                    <Text style={styles.categoryHeader}>{title}</Text>
                    <RNPickerSelect
                        onValueChange={(value) => updateCategorySelected(value)}
                        items={selectItems}
                        placeholder={{
                            label: 'Select a Category',
                            value: null,
                            color: AppConstants.colors.black
                        }}
                        style={{
                            viewContainer: { 
                                justifyContent: 'center',
                                width: '100%',
                                flex: 1
                            },
                            inputAndroid: {
                                fontSize: normalize(15),
                                color: AppConstants.colors.black
                            },
                            inputIOS: {
                                fontSize: normalize(15),
                                color: AppConstants.colors.black,
                                textAlign: 'right'
                            },
                            placeholder: {
                                fontSize: normalize(15),
                                color: AppConstants.colors.black
                            }
                        }}
                    />
                </View>
                <FlatList
                    horizontal
                    data={mediaList}
                    renderItem={(item) => renderMediaListItem(item, categoriesList, categorySelected)}
                    showsHorizontalScrollIndicator={false}
                />
            </View>
        )
    }

    const renderMediaListItem = ({ item }: any, categoriesList: Array<Category>, categorySelected: number) => {

        const filteredArray = categoriesList.filter(value => item.categories.includes(value.id));
        let categories = filteredArray.map((item) => item.genre).join(', ')


        return (
            _.isNull(categorySelected) || item.categories.includes(categorySelected) ?
                <MediaListItem
                    item={item}
                    onPress={viewItemDetails}
                    categories={categories}
                /> : null
        )
    }


    return (
        <SafeAreaView style={styles.centeredView}>
            <View style={styles.container}>
                <SearchBar
                    placeholder="Discover"
                    onChangeText={doSearch}
                    value={search}
                    containerStyle={[commonStyles.shadows, styles.searchContainer, { zIndex: -99 }]}
                    inputContainerStyle={{ backgroundColor: AppConstants.colors.white, zIndex: -99 }}
                />
                <ScrollView
                    scrollEnabled={true}
                    showsVerticalScrollIndicator={false}
                >
                    {!_.isEmpty(movies) &&
                        renderPrincipalMovies()
                    }
                    {!_.isEmpty(moviesCategories) &&
                        renderCategoryItems(moviesCategories, "Movies", setMovieCategorySelected, movies, movieCategorySelected)
                    }
                    {!_.isEmpty(tvSeriesCategories) &&
                        renderCategoryItems(tvSeriesCategories, "TV Series", setTvSerieCategorySelected, tvSeries, tvSerieCategorySelected)
                    }
                </ScrollView>
            </View>
        </SafeAreaView>
    );
}

export default IsLoadingHoc(HomeScreen);


const styles = StyleSheet.create({
    centeredView: {
        backgroundColor: AppConstants.colors.principalColor,
        flex: 1,
        height: '100%',
        alignItems: 'center',
        justifyContent: 'center',
    },
    container: {
        height: '100%',
        width: '100%',
        padding: normalize(10),
    },
    searchContainer: {
        padding: 5,
        backgroundColor: AppConstants.colors.white,
        borderTopWidth: 0,
        borderBottomWidth: 0,
        borderRadius: 5,
        width: '100%',
        marginBottom: normalize(30)
    },
    dropDownStyle: {
        width: '100%',
        backgroundColor: 'transparent',
        flex: 2
    },
    dropDownText: {
        textAlign: 'right'
    },
    categoriesContainer: {
        marginTop: normalize(20),
        marginBottom: normalize(10),
        flexDirection: 'row',
    },
    categoryHeader: {
        fontSize: normalize(25),
        alignSelf: 'center'
    }
});
