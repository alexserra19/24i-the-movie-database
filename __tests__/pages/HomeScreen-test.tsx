import 'react-native';
import React from 'react';
import HomeScreen from '../../src/pages/HomeScreen';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store'
import { Store, AnyAction } from 'redux';
import { render } from '@testing-library/react-native'
import { Dimensions } from 'react-native';

jest.useFakeTimers()

describe('HomeScreen', () => {

  let store: Store<any, AnyAction>
  const mockStore = configureStore()

  const initialState = {
    mediaReducer: {
      selectedMedia: {}
    }
  }

  beforeEach(() => {
    store = mockStore(initialState)
  });

  const mockProps = {
    navigation: {},
    setLoading: jest.fn()
  }

  it('should match the snapshot', () => {
    const wrapper = render(
      <Provider store={store}>
        <HomeScreen {...mockProps} />
      </Provider>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  })

  it('useEffect', () => {
    jest.spyOn(Dimensions, 'addEventListener')
    const wrapper = render(
      <Provider store={store}>
        <HomeScreen {...mockProps} />
      </Provider>
    );
    expect(Dimensions.addEventListener).toBeCalled()
  })

});
