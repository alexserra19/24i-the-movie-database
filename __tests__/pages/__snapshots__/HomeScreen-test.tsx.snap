// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen should match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E1E2E6",
      "flex": 1,
      "height": "100%",
      "justifyContent": "center",
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "height": "100%",
        "padding": 12.857142857142858,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffff",
          "borderBottomColor": "#000",
          "borderBottomWidth": 0,
          "borderRadius": 5,
          "borderTopColor": "#000",
          "borderTopWidth": 0,
          "elevation": 9,
          "marginBottom": 38.57142857142857,
          "padding": 5,
          "shadowColor": "#000",
          "shadowOffset": Object {
            "height": 4,
            "width": 0,
          },
          "shadowOpacity": 0.32,
          "shadowRadius": 5.46,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "paddingHorizontal": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#ffff",
              "borderBottomWidth": 0,
              "borderColor": "#86939e",
              "borderRadius": 3,
              "flexDirection": "row",
              "minHeight": 30,
              "overflow": "hidden",
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginLeft": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "overflow": "hidden",
                }
              }
            >
              <View>
                <View
                  style={
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "transparent",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Text
                    allowFontScaling={false}
                    style={
                      Array [
                        Object {
                          "color": "#86939e",
                          "fontSize": 18,
                        },
                        Object {
                          "backgroundColor": "transparent",
                        },
                        Object {
                          "fontFamily": "Material Icons",
                          "fontStyle": "normal",
                          "fontWeight": "normal",
                        },
                        Object {},
                      ]
                    }
                    testID="iconIcon"
                  >
                    î¢¶
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <TextInput
            allowFontScaling={true}
            editable={true}
            onBlur={[Function]}
            onCancel={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="Discover"
            placeholderTextColor="#86939e"
            platform="default"
            rejectResponderTermination={true}
            style={
              Object {
                "color": "#86939e",
                "flex": 1,
                "fontSize": 18,
                "marginLeft": 10,
                "minHeight": 40,
              }
            }
            testID="searchInput"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "marginRight": 8,
                "marginVertical": 4,
                "paddingRight": 4,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                }
              }
            />
          </View>
        </View>
        <Text
          style={
            Object {
              "color": "#ff190c",
              "fontSize": 12,
              "height": 0,
              "margin": 0,
              "padding": 0,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "bottom": 0,
            "elevation": 9,
            "flex": 1,
            "justifyContent": "center",
            "left": 0,
            "opacity": 0.4,
            "position": "absolute",
            "right": 0,
            "top": 0,
            "zIndex": 99,
          },
          Object {
            "backgroundColor": "transparent",
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#000000"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
    <RCTScrollView
      contentContainerStyle={
        Object {
          "alignContent": "center",
        }
      }
      scrollEnabled={true}
      showsVerticalScrollIndicator={false}
    >
      <View />
    </RCTScrollView>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#ffff",
            "borderRadius": 10,
            "padding": 45,
            "width": 385.7142857142857,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 25.714285714285715,
              "fontWeight": "bold",
              "marginBottom": 32.14285714285714,
              "textAlign": "center",
            }
          }
        >
          No Results Found
        </Text>
        <Text
          style={
            Object {
              "fontSize": 19.285714285714285,
              "marginBottom": 25.714285714285715,
              "textAlign": "center",
            }
          }
        >
          Hmmmm, we're not getting any results. Our bad, try another search
        </Text>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "#409ff7",
              "borderRadius": 10,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Ok
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</RCTSafeAreaView>
`;
